crates/matrix-sdk/src/authentication/qrcode/mod.rs

```diff
@@ -0,0 +1,216 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Types for the QR code login support defined in [MSC4108](https://github.com/matrix-org/matrix-spec-proposals/pull/4108).
+//!
+//! Please note, QR code logins are only supported when using OIDC as the
+//! auththentication mechanism, native Matrix authentication does not support
+//! it.
+//!
+//! This currently only implements the case where the new device is scanning the
+//! QR code. To log in using a QR code, please take a look at the
+//! [`Oidc::login_with_qr_code()`] method
+
+use as_variant::as_variant;
+use matrix_sdk_base::crypto::SecretImportError;
+pub use openidconnect::{
+    core::CoreErrorResponseType, ConfigurationError, DeviceCodeErrorResponseType, DiscoveryError,
+    HttpClientError, RequestTokenError, StandardErrorResponse,
+};
+use thiserror::Error;
+use url::Url;
+pub use vodozemac::ecies::{Error as EciesError, MessageDecodeError};
+
+#[cfg(doc)]
+use crate::oidc::Oidc;
+use crate::{oidc::CrossProcessRefreshLockError, HttpError};
+
+mod login;
+mod messages;
+mod oidc_client;
+mod rendezvous_channel;
+mod secure_channel;
+
+pub use matrix_sdk_base::crypto::types::qr_login::{
+    LoginQrCodeDecodeError, QrCodeData, QrCodeMode, QrCodeModeData,
+};
+
+pub use self::{
+    login::{LoginProgress, LoginWithQrCode},
+    messages::{LoginFailureReason, QrAuthMessage},
+};
+
+/// The error type for failures while trying to log in a new device using a QR
+/// code.
+#[derive(Debug, Error)]
+#[cfg_attr(feature = "uniffi", derive(uniffi::Error), uniffi(flat_error))]
+pub enum QRCodeLoginError {
+    /// An error happened while we were communicating with the OIDC provider.
+    #[error(transparent)]
+    Oidc(#[from] DeviceAuhorizationOidcError),
+
+    /// The other device has signaled to us that the login has failed.
+    #[error("The login failed, reason: {reason}.")]
```

Comment:
I think it's better to avoid `.` at the end of the error message. Especially if the reason already contains a `.`.

Alternatively, you can put the reason between quotes, like so:

```diff
- "The login failed, reason: {reason}."
+ "The login failed, reason: “{reason}”."
```

Reply:
Removed them dots in 6a66cdf420f2324d40bc6adc240c9e6a11778869.